### PR TITLE
FilledArea Annotation

### PR DESCRIPTION
**Pull request by canavandl**
Part of #2352

- [x] tests added / passed
- [ ] release document entry (if new feature or API change)

### CODE REVIEW

**Comment by canavandl** [comment]
initial output:

screen shot 2017-04-20 at 4 28 18 pm

```python
from bokeh.plotting import figure, show
from bokeh.models import FilledArea, ColumnDataSource

p = figure(plot_width=400, plot_height=400, tools='pan')
source = ColumnDataSource(data=dict(
    x=[1,2,3,4,5],
    y=[4,2,4,5,3],
    y_upper=[5,3,5,6,4],
    y_lower=[3,1,3,4,2]
))

p.line(x='x', y='y', source=source)

area = FilledArea(
    coords='x', dimension='height', lower='y_lower', upper='y_upper', source=source,
    level='underlay'
)
p.add_layout(area)

show(p)
```

File: bokeh/models/annotations.py [hunk 1]
```diff
@@ -530,6 +530,52 @@ class BoxAnnotation(Annotation):
...
+    '''
+    lower = NumberSpec(help="""
+    The lower coordinates to locate the text anchors.
+    """)
+
+    upper = NumberSpec(help="""
+    The upper coordinates to locate the text anchors.
+    """)
+
+    coords = NumberSpec(help="""
```
Comment (canavandl) [inline bokeh/models/annotations.py]: What's a better name for this property?

**Comment by canavandl** [comment]
Questions:
* Do we need to support screen coordinates in addition to data coords?
* Are we happy with the upper/lower/dimension coordinate system instead of (lower_x, lower_y, upper_x, upper_y?)

**Comment by bryevdv** [comment]
I think `upper` and `lower` is great. Do we want to add an orientation for filled areas across the horizontal dimension?

File: bokeh/models/annotations.py (hunk 1 above)
Comment (bryevdv) [inline bokeh/models/annotations.py]: `base`?

**Comment by canavandl** [comment]
> I think upper and lower is great. Do we want to add an orientation for filled areas across the horizontal dimension?
If I'm understanding your question correctly, that's what the `dimension` property was for. Though I realize I should be using 'orientation' instead.

File: bokeh/models/annotations.py (hunk 1 above)
Comment (canavandl) [inline bokeh/models/annotations.py]: base! That'll work

Comment (bryevdv) [inline bokeh/models/annotations.py]: is this necessary though? I think we just have users provide upper and lower coords as absolute coords, and if they want to tie them to some other base set of coords a transform will work for that.

**Comment by bryevdv** [comment]
`Band` might also be a good but shorter name, just a thought.

**Comment by bryevdv** [comment]
> If I'm understanding your question correctly, that's what the dimension property was for. Though I realize I should be using 'orientation' instead. 

Sorry I just missed that, `dimension` is fine.

File: bokeh/models/annotations.py (hunk 1 above)
Comment (bryevdv) [inline bokeh/models/annotations.py]: Don't have to worry about not but worth considering the case where upper and lower are different lengths... I think that implies multiple sources but may theres just a new different annotation for that.

**Comment by canavandl** [comment]
I started writing writing JS unit tests, but there's so little testable surface in the view (most of the code is canvas drawing) and nothing in the model. I'll probably just add some selenium tests instead.

**Comment by canavandl** [comment]
Current status:

screen shot 2017-04-21 at 10 33 05 am

I think it'd look better without the lines on the outer sides.

**Comment by bryevdv** [comment]
> I think it'd look better without the lines on the outer sides.
  
Agree

File: bokehjs/src/coffee/models/annotations/band.coffee [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+  set_data: (source) ->
+    super(source)
+    @visuals.warm_cache(source)
+
+  _map_data: () ->
+    x_mapper = @plot_view.frame.x_mappers[@model.x_range_name]
+    y_mapper = @plot_view.frame.y_mappers[@model.y_range_name]
+
+    non_base_mapper = if @model.dimension == "height" then y_mapper else x_mapper
+    base_mapper = if @model.dimension == "height" then x_mapper else y_mapper
```
Comment (bryevdv) [inline bokehjs/src/coffee/models/annotations/band.coffee]: maybe `base_mapper` and `limit_mapper`?

File: bokeh/models/annotations.py [hunk 3]
```diff
@@ -530,6 +530,71 @@ class BoxAnnotation(Annotation):
...
+    ''' Render a filled area band along a dimension.
+
+    '''
+    lower = NumberSpec(help="""
+    The coordinates of the lower portion of the filled area band.
+    """)
+
+    lower_units = Enum(SpatialUnits, default='data', help="""
+    The unit type for the lower attribute. Interpreted as "data space" units
+    by default.
```
Comment (canavandl) [inline bokeh/models/annotations.py]: Fixed

File: bokehjs/src/coffee/models/annotations/band.coffee (hunk 2 above)
Comment (canavandl) [inline bokehjs/src/coffee/models/annotations/band.coffee]: Fixed

**Comment by canavandl** [comment]
Here's the User Guide example:

screen shot 2017-04-24 at 10 58 07 am

**Comment by canavandl** [comment]
Related to the above example, it seems useful to also implement a ``center`` or ``middle`` line (perhaps with unique styling), so that it's possible to indicate the rolling mean.

Does that seem reasonable?

**Comment by bryevdv** [comment]
> Related to the above example, it seems useful to also implement a center or middle line (perhaps with unique styling), so that it's possible to indicate the rolling mean.

> Does that seem reasonable?

I don't think we should be in the business of computing means ourselves, and there are already other ways to draw lines so I don't see the benefit of extra complexity

**Comment by canavandl** [comment]
> I don't think we should be in the business of computing means ourselves, and there are already other ways to draw lines so I don't see the benefit of extra complexity

That's fine. I'll mark this PR as ready then.

**Comment by canavandl** [comment]
I forgot to add the TS Typings.

**Comment by mattpap** [comment]
120 changed files? Something is pretty wrong with this PR.

**Comment by canavandl** [comment]
Nah - I just messed up a rebase.

**Comment by StevenCHowell** [comment]
@canavandl this looks nice!  I was just wanting to generate a plot like this and wondered if quad was still the best option.  Looking forward to using this!

**Pull request merged**